lib/diaspora/federation/dispatcher.rb

```diff
@@ -0,0 +1,84 @@
+module Diaspora
+  module Federation
+    class Dispatcher
+      include Diaspora::Logging
+
+      def initialize(sender, object, opts={})
+        @sender = sender
+        @object = object
+        @opts = opts
+      end
+
+      def self.build(sender, object, opts={})
+        if object_should_be_processed_as_public?(object)
+          Public.new(sender, object, opts)
+        else
+          Private.new(sender, object, opts)
+        end
+      end
+
+      def self.defer_dispatch(sender, object, opts={})
+        object.save if object.new_record?
+        Workers::DeferredDispatch.perform_async(sender.id, object.class.to_s, object.id, opts)
+      end
+
+      def dispatch
+        deliver_to_services
+        deliver_to_subscribers
+      end
+
+      private
+
+      attr_reader :sender, :object, :opts
+
+      def self.object_should_be_processed_as_public?(object)
+        object.respond_to?(:public?) && object.public?
+      end
+      private_class_method :object_should_be_processed_as_public?
+
+      def deliver_to_services
+        deliver_to_user_services if opts[:service_types]
+      end
+
+      def deliver_to_subscribers
+        local_people, remote_people = subscribers.partition(&:local?)
+
+        deliver_to_local(local_people) unless local_people.empty?
+        deliver_to_remote(remote_people)
+      end
+
+      def deliver_to_local(people)
+        obj = object.respond_to?(:object_to_receive) ? object.object_to_receive : object
```

Comment:
this doesn't work, because `object_to_receive` can return nil, and then it shouldn't use `object`.